packages/editor/src/components/post-locked-modal/index.js

```diff
@@ -0,0 +1,230 @@
+
+/**
+ * Internal dependencies
+ */
+import { getWPAdminURL } from '../../utils/url';
+import PostPreviewButton from '../post-preview-button';
+
+class PostLockedModal extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.sendPostLock = this.sendPostLock.bind( this );
+		this.receivePostLock = this.receivePostLock.bind( this );
+		this.releasePostLock = this.releasePostLock.bind( this );
+	}
+
+	componentDidMount() {
+		// Details on these events on the Heartbeat API docs
+		// https://developer.wordpress.org/plugins/javascript/heartbeat-api/
+		jQuery( document )
+			.on( 'heartbeat-send.refresh-lock', this.sendPostLock )
+			.on( 'heartbeat-tick.refresh-lock', this.receivePostLock );
+	}
+
+	componentWillUnmount() {
+		jQuery( document )
+			.off( 'heartbeat-send.refresh-lock', this.sendPostLock )
+			.off( 'heartbeat-tick.refresh-lock', this.receivePostLock );
+	}
+
+	/**
+	 * Keep the lock refreshed.
+	 *
+	 * When the user does not send a heartbeat in a heartbeat-tick
+	 * the user is no longer editing and another user can start editing.
+	 *
+	 * @param {Object} event Event.
+	 * @param {Object} data  Data to send in the heartbeat request.
+	 */
+	sendPostLock( event, data ) {
+		const { isLocked, activePostLock, postId } = this.props;
+		if ( isLocked ) {
+			return;
+		}
+
+		data[ 'wp-refresh-post-lock' ] = {
+			lock: activePostLock,
+			post_id: postId,
+		};
+	}
+
+	/**
+	 * Refresh post locks: update the lock string or show the dialog if somebody has taken over editing.
+	 *
+	 * @param {Object} event Event.
+	 * @param {Object} data  Data received in the heartbeat request
+	 */
+	receivePostLock( event, data ) {
+		if ( ! data[ 'wp-refresh-post-lock' ] ) {
+			return;
+		}
+
+		const { autosave, updatePostLock } = this.props;
+		const received = data[ 'wp-refresh-post-lock' ];
+		if ( received.lock_error ) {
+			// Auto save and display the takeover modal.
+			autosave();
+			updatePostLock( {
+				isLocked: true,
+				isTakeover: true,
+				user: {
+					avatar: received.lock_error.avatar_src,
+				},
+			} );
+		} else if ( received.new_lock ) {
+			updatePostLock( {
+				isLocked: false,
+				activePostLock: received.new_lock,
+			} );
+		}
+	}
+
+	/**
+	 * Unlock the post before the window is exited.
+	 *
+	 * @param {Object} event Event.
+	 */
+	releasePostLock() {
+		const { isLocked, activePostLock, postLockUtils, postId } = this.props;
+		if ( isLocked || ! activePostLock ) {
+			return;
+		}
+
+		const data = {
+			action: 'wp-remove-post-lock',
+			_wpnonce: postLockUtils.unlockNonce,
+			post_ID: postId,
+			active_post_lock: activePostLock,
+		};
+
+		jQuery.post( {
+			async: false,
+			url: postLockUtils.ajaxUrl,
+			data,
+		} );
+	}
+
+	render() {
+		const { user, postId, isLocked, isTakeover, postLockUtils, postType } = this.props;
+		if ( ! isLocked ) {
+			return null;
+		}
+
+		const userDisplayName = user.name || __( 'Another user' );
```

Comment:
Unsure if this could use a translator note, since it's a fragment of a sentence that lacks context.